google-cloud-storage/lib/google/cloud/storage/hmac_key.rb

```diff
@@ -0,0 +1,309 @@
+        # Whether the state of the HMAC key is `INACTIVE`.
+        #
+        # @return [Boolean]
+        #
+        def inactive?
+          state == "INACTIVE"
+        end
+
+        ##
+        # Updates the state of the HMAC key to `INACTIVE`.
+        #
+        # @return [Boolean] Returns `true` if the HMAC key was updated.
+        #
+        # @example
+        #   require "google/cloud/storage"
+        #
+        #   storage = Google::Cloud::Storage.new
+        #
+        #   hmac_key = storage.hmac_keys.first
+        #
+        #   hmac_key.inactive!
+        #   hmac_key.state # "INACTIVE"
+        #
+        def inactive!
+          put_gapi! "INACTIVE"
+          self
+        end
+
+        ##
+        # Whether the state of the HMAC key is `DELETED`.
+        #
+        # @return [Boolean]
+        #
+        def deleted?
+          state == "DELETED"
+        end
+
+        ##
+        # Deletes the HMAC key, and loads the new state of the HMAC key,
+        # which will be `DELETED`.
+        #
+        # The API call to delete the notification may be retried under certain
+        # conditions. See {Google::Cloud#storage} to control this behavior.
+        #
+        # @return [Boolean] Returns `true` if the notification was deleted.
+        #
+        # @example
+        #   require "google/cloud/storage"
+        #
+        #   storage = Google::Cloud::Storage.new
+        #
+        #   hmac_key = storage.hmac_keys.first
+        #
+        #   hmac_key.inactive!.delete!
+        #
+        def delete!
+          ensure_service!
+          @service.delete_hmac_key access_id,
+                                   user_project: @user_project
+          @gapi = @service.get_hmac_key access_id,
+                                           user_project: @user_project
+          self
+        end
+        alias delete delete!
+
+        ##
+        # Reloads the HMAC key with current data from the Storage service.
+        #
+        def reload!
+          ensure_service!
+          @gapi = service.get_hmac_key access_id, user_project: user_project
+          self
+        end
+        alias refresh! reload!
+
+        ##
+        # @private New HmacKey from a Google::Apis::StorageV1::HmacKey object.
+        #
+        # @param [Google::Apis::StorageV1::HmacKey] gapi
+        #
+        #
+        def self.from_gapi gapi, service, user_project: nil
+          hmac_key = self.from_gapi_metadata gapi.metadata,
+                                             service,
+                                             user_project: user_project
+          hmac_key.tap do |f|
+            f.instance_variable_set :@secret, gapi.secret
+          end
+        end
+
+        ##
+        # @private New HmacKey from a Google::Apis::StorageV1::HmacKeyMetadata
+        # object.
+        #
+        # @param [Google::Apis::StorageV1::HmacKeyMetadata] gapi
+        #
+        #
+        def self.from_gapi_metadata gapi, service, user_project: nil
```

Comment:
@blowmage Are you OK with this second "mode" when the instance is initialized only with the `Google::Apis::StorageV1::HmacKeyMetadata` data? This is how it is presented by every operation except create, which uses the full representation (`from_gapi`, above).

Reply:
Only one field is missing in this mode, the HMAC `secret`.

Reply:
As long as it's private and we can refactor it without breaking the public API, so I have no objections.